src/server.rs

```diff
@@ -12,48 +29,207 @@ pub struct ServerConfig {
+    let key = tokio::fs::read(&config.server.server_key_path).await?;
+    let cert: Vec<rustls::Certificate> = rustls_pemfile::certs(&mut cert.as_slice())
+        .unwrap()
+        .iter()
+        .map(|v| rustls::Certificate(v.clone()))
+        .collect();
+    let key = match rustls_pemfile::read_one(&mut key.as_slice())
+        .expect("cannot parse private key .pem file")
+    {
+        Some(rustls_pemfile::Item::RSAKey(key)) => rustls::PrivateKey(key),
+        Some(rustls_pemfile::Item::PKCS8Key(key)) => rustls::PrivateKey(key),
+        Some(rustls_pemfile::Item::ECKey(key)) => rustls::PrivateKey(key),
+        _ => panic!("cannot parse private key .pem file"),
+    };
+    let mut server_config = rustls::ServerConfig::new(client_auth);
+    server_config.set_single_cert(cert, key).unwrap();
+
+    Ok(server_config)
+}
+
+pub(crate) fn socket_addr(config: &Arc<StewardConfig>) -> Result<SocketAddr, Error> {
     let port = &config.server.port;
     let address = &config.server.address;
     let address: SocketAddr = format!("{}:{}", address, port).parse()?;
 
+    Ok(address)
+}
+
+pub async fn load_server_config(config: &Arc<StewardConfig>) -> Result<ServerConfig, Error> {
+    let server_config = auth_config(config).await?;
+    let tls_config = ServerTlsConfig::new()
+        .rustls_server_config(server_config)
+        .to_owned();
+    let address = socket_addr(config)?;
+
     Ok(ServerConfig {
         tls_config: Some(tls_config),
         address,
     })
 }
 
-pub async fn load_tls_config(
-    server_cert_path: &str,
-    server_key_path: &str,
-    client_ca_path: Option<String>,
-) -> Result<ServerTlsConfig, Error> {
-    let cert = tokio::fs::read(server_cert_path).await?;
-    let key = tokio::fs::read(server_key_path).await?;
-    let server_identity = Identity::from_pem(cert, key);
-    let client_ca = match client_ca_path {
-        Some(path) => tokio::fs::read(path).await?,
-        None => DEFAULT_CLIENT_CA.into(),
-    };
-    let client_ca_cert = Certificate::from_pem(client_ca);
+/// A client is authorized to publish to a subscription if the following criteria are met:
+/// 1. The client certificate has an authority key identifier extension whose value is the key of an entry
+///    in the publisher trust state cache
+/// 2. The client certificate is signed by the publisher's CA certificate
+/// 3. The entry's list of subscription IDs contains the subscription ID of the request
+pub(crate) async fn handle_authorization(
+    subscription_id: &String,
+    certs: Arc<Vec<Certificate>>,
+) -> Result<(), Status> {
+    // In case the request contains multiple certificates, we iterate until the first valid one.
+    for cert in certs.iter() {
+        let (_, client_cert) = match X509Certificate::from_der(cert.get_ref()) {
+            Ok(cert) => cert,
+            Err(e) => {
+                error!("failed to parse certificate for authorization check: {e}");
+                continue;
```

Comment:
The assumption was if one is valid then it will get all the way to `Ok(())` and we return. If none, the loop gets exited by the last `continue` and the error at the end of the function gets returned. 

I think this structure is going to change given your comment on the need to find the client cert in the chain.

Reply:
We'll come back to this after resolving the other comment.